streamingvisitors/src/tests/matching_elements_filler/matching_elements_filler_test.cpp

```diff
@@ -0,0 +1,422 @@
+    std::unique_ptr<MapFieldValue> make_str_int_map(std::map<std::string, int> values) const;
+    FieldPath make_field_path(vespalib::string path) const;
+    std::unique_ptr<document::Document> make_test_doc() const;
+};
+
+MyDocType::MyDocType()
+    : _name_field("name", 1, *DataType::STRING, true),
+      _weight_field("weight", 2, *DataType::INT, true),
+      _elem_type(make_elem_type(_name_field, _weight_field)),
+      _elem_array_type(_elem_type),
+      _elem_map_type(*DataType::STRING, _elem_type),
+      _str_int_map_type(*DataType::STRING, *DataType::INT),
+      _elem_array_field("elem_array", 3, _elem_array_type, true),
+      _elem_map_field("elem_map", 4, _elem_map_type, true),
+      _str_int_map_field("str_int_map", _str_int_map_type, true),
+      _document_type("test")
+{
+    _document_type.addField(_elem_array_field);
+    _document_type.addField(_elem_map_field);
+    _document_type.addField(_str_int_map_field);
+}
+
+MyDocType::~MyDocType() = default;
+
+std::unique_ptr<StructFieldValue>
+MyDocType::make_elem(const vespalib::string& name, int weight) const
+{
+    auto ret = std::make_unique<StructFieldValue>(_elem_type);
+    ret->setValue(_name_field, StringFieldValue(name));
+    ret->setValue(_weight_field, IntFieldValue(weight));
+    return ret;
+}
+
+std::unique_ptr<ArrayFieldValue>
+MyDocType::make_elem_array(std::vector<std::pair<std::string, int>> values) const
+{
+    auto ret = std::make_unique<ArrayFieldValue>(_elem_array_type);
+    for (auto& name_weight : values) {
+        ret->add(*make_elem(name_weight.first, name_weight.second));
+    }
+    return ret;
+}
+
+std::unique_ptr<MapFieldValue>
+MyDocType::make_elem_map(std::map<std::string, std::pair<std::string, int>> values) const
+{
+    auto ret = std::make_unique<MapFieldValue>(_elem_map_type);
+    for (auto& kv : values) {
+        ret->put(StringFieldValue(kv.first), *make_elem(kv.second.first, kv.second.second));
+    }
+    return ret;
+}
+
+std::unique_ptr<MapFieldValue>
+MyDocType::make_str_int_map(std::map<std::string, int> values) const
+{
+    auto ret = std::make_unique<MapFieldValue>(_str_int_map_type);
+    for (auto& kv : values) {
+        ret->put(StringFieldValue(kv.first), IntFieldValue(kv.second));
+    }
+    return ret;
+}
+
+FieldPath
+MyDocType::make_field_path(vespalib::string path) const
+{
+    FieldPath ret;
+    _document_type.buildFieldPath(ret, path);
+    return ret;
+}
+
+std::unique_ptr<document::Document>
+MyDocType::make_test_doc() const
+{
+    auto doc = std::make_unique<document::Document>(_document_type, DocumentId("id::test::1"));
+    doc->setValue("elem_array", *make_elem_array({{"foo", 10},{"bar", 20},{"baz", 30},{"foo", 40}, {"zap", 20}, {"zap", 20}}));
+    doc->setValue("elem_map", *make_elem_map({{"@foo", {"foo", 10}}, {"@bar", {"bar", 20}},{"@baz", {"baz", 30}},{"@foo@", {"foo", 40}},{"@zap", {"zap", 20}}, {"@zap@", {"zap", 20}}}));
```

Comment:
Consider adding a note that the elements in maps are ordered on the key, otherwise it is hard to understand the expected results in the tests.